src/Plutus/Types/Value.purs

```diff
@@ -0,0 +1,227 @@
+unionWith f lhs =
+  Value <<< map (map (these identity identity f)) <<< unionVal lhs
+
+-- https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/src/Plutus.V1.Ledger.Value.html#flattenValue
+-- | Convert a value to a simple list, keeping only the non-zero amounts.
+flattenValue :: Value -> Array (CurrencySymbol /\ TokenName /\ BigInt)
+flattenValue (Value (Plutus.Map arr)) =
+  flip concatMap arr \(cs /\ (Plutus.Map tokens)) ->
+    tokens <#> \(tn /\ value) ->
+      cs /\ tn /\ value
+
+-- https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/src/Plutus.V1.Ledger.Value.html#geq
+-- | Check whether one `Value` is greater than or equal to another.
+geq :: Value -> Value -> Boolean
+geq = checkBinRel (>=)
+
+-- https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/src/Plutus.V1.Ledger.Value.html#gt
+-- | Check whether one `Value` is strictly greater than another.
+gt :: Value -> Value -> Boolean
+gt l r = not (isZero l && isZero r) && checkBinRel (>) l r
+
+-- https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/src/Plutus.V1.Ledger.Value.html#leq
+-- | Check whether one `Value` is less than or equal to another.
+leq :: Value -> Value -> Boolean
+leq = checkBinRel (<=)
+
+-- https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/src/Plutus.V1.Ledger.Value.html#lt
+-- | Check whether one `Value` is strictly less than another.
+lt :: Value -> Value -> Boolean
+lt l r = not (isZero l && isZero r) && checkBinRel (<) l r
+
+-- https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/src/Plutus.V1.Ledger.Value.html#eq
+-- | Check whether one `Value` is equal to another.
+eq :: Value -> Value -> Boolean
+eq = checkBinRel (==)
+
+--------------------------------------------------------------------------------
+-- Internal Functions
+--------------------------------------------------------------------------------
+
+-- https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/src/Plutus.V1.Ledger.Value.html#unionVal
+-- Combines two 'Value' maps.
+unionVal
+  :: Value
+  -> Value
+  -> Plutus.Map CurrencySymbol (Plutus.Map TokenName (These BigInt BigInt))
+unionVal (Value lhs') (Value rhs) =
+  these (map This) (map That) Plutus.Map.union <$>
+    Plutus.Map.union lhs' rhs
+
+-- https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/src/Plutus.V1.Ledger.Value.html#checkPred
+checkPred :: (These BigInt BigInt -> Boolean) -> Value -> Value -> Boolean
+checkPred f l r = all (all f) (unionVal l r)
+
+-- https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/src/Plutus.V1.Ledger.Value.html#checkBinRel
+-- Check whether a binary relation holds for value pairs of two `Value` maps,
+-- supplying 0 where a key is only present in one of them.
+checkBinRel :: (BigInt -> BigInt -> Boolean) -> Value -> Value -> Boolean
+checkBinRel f l r = checkPred (these (f zero) (f zero) f) l r
```

Comment:
I'm curious why are there `zero`s. As a result of them, comparison functions do treat negative values differently when they are present only in one of the values.

I.e. `( A -> B1 -> -1) < (A -> B2 -> 1)` is `false` which is counter-intuitive. Code to reproduce:

```
        let b1 = hexToByteArrayUnsafe ""
        let ada = unsafePartial $ fromJust $ V.mkCurrencySymbol b1
        let tn1 = unsafePartial $ fromJust $ mkTokenName $ hexToByteArrayUnsafe "00"
        let tn2 = unsafePartial $ fromJust $ mkTokenName $ hexToByteArrayUnsafe "11"
        let v1 = V.singleton ada tn1 $ negate one
        let v2 = V.singleton ada tn2 $ one
        liftEffect $ Console.log $ show $ Plutus.Types.Value.lt v1 v2
```

(This is not a change request since it's copied from upstream).

Reply:
@klntsky Thanks for pointing that out! 
There was actually a pretty obvious bug, which caused this weird behaviour. 

It seems to have been fixed after passing `flip f zero` as the first parameter to `these`. Thus, if some amount is present only in the lhs value, it is compared to `zero` and not the other way around.

```PureScript
checkBinRel :: (BigInt -> BigInt -> Boolean) -> Value -> Value -> Boolean
checkBinRel f l r = checkPred (these (flip f zero) (f zero) f) l r
```